cabal-install/src/Distribution/Client/ScriptUtils.hs

```diff
@@ -0,0 +1,354 @@
+  -> IO b
+withContextAndSelectors noTargets cachePrefix kind flags@NixStyleFlags {..} targetStrings globalFlags act
+  = withTemporaryTempDirectory $ \mkTmpDir -> do
+    (tc, ctx) <- withProjectOrGlobalConfig verbosity ignoreProject globalConfigFlag with (without mkTmpDir)
+
+    -- In the case where a selector is both a valid target and script, assume it is a target,
+    -- because you can disambiguate the script with "./script"
+    let truncateExe = if kind == Just ExeKind then take 1 else id
+    (tc', ctx', sels) <- readTargetSelectors (localPackages ctx) kind (truncateExe targetStrings) >>= \case
+      Left err@(TargetSelectorNoTargetsInProject:_)
+        | [] <- targetStrings
+        , AcceptNoTargets <- noTargets -> return (tc, ctx, defaultTarget)
+        | (script:_) <- targetStrings  -> scriptOrError script err
+      Left err@(TargetSelectorNoSuch t _:_)
+        | TargetString1 script <- t    -> scriptOrError script err
+      Left err@(TargetSelectorExpected t _ _:_)
+        | TargetString1 script <- t    -> scriptOrError script err
+      Left err                         -> reportTargetSelectorProblems verbosity err
+      Right sels                       -> return (tc, ctx, sels)
+
+    act tc' ctx' sels
+  where
+    verbosity = fromFlagOrDefault normal (configVerbosity configFlags)
+    ignoreProject = flagIgnoreProject projectFlags
+    cliConfig = commandLineFlagsToProjectConfig globalFlags flags mempty
+    globalConfigFlag = projectConfigConfigFile (projectConfigShared cliConfig)
+    defaultTarget = [TargetPackage TargetExplicitNamed [fakePackageId] Nothing]
+
+    with = do
+      ctx <- establishProjectBaseContext verbosity cliConfig OtherCommand
+      return (ProjectContext, ctx)
+    without mkDir globalConfig = do
+      distDirLayout <- establishDummyDistDirLayout verbosity (globalConfig <> cliConfig) =<< mkDir
+      ctx <- establishDummyProjectBaseContext verbosity (globalConfig <> cliConfig) distDirLayout [] OtherCommand
+      return (GlobalContext, ctx)
+    scriptOrError script err = do
+      exists <- doesFileExist script
+      if exists then do
+        -- In the script case we always want a dummy context even when ignoreProject is False
+        let mkCacheDir = ensureScriptCacheDirectory verbosity cachePrefix script
+        (_, ctx) <- withProjectOrGlobalConfig verbosity (Flag True) globalConfigFlag with (without mkCacheDir)
+
+        let projectRoot = distProjectRootDirectory $ distDirLayout ctx
+        writeFile (projectRoot </> "scriptlocation") =<< canonicalizePath script
+
+        (executable, contents) <- readScriptBlockFromScript verbosity =<< BS.readFile script
+
+        let executable' = executable & L.buildInfo . L.defaultLanguage %~ maybe (Just Haskell2010) Just
+        return (ScriptContext script executable' contents, ctx, defaultTarget)
+      else reportTargetSelectorProblems verbosity err
+
+withTemporaryTempDirectory :: (IO FilePath -> IO a) -> IO a
```

Comment:
Just in case: have you considered use https://github.com/haskell/cabal/blob/6a8c3f80173b71436386d9724b64eb67ccb1cf40/Cabal/src/Distribution/Simple/Utils.hs#L1467-L1470?

Reply:
It's a temporary temp directory, that's twice as temporary! Being serious though, it does use `withTempDirectory`. With `withTempDirectory` creates a directory in a given dir using the template meant to me used as a temp directory. `withTemporaryTempDirectory` does that under /tmp or os equivilent, hence temporary temp.